CSS Module: basic UI (#24152)
mdn/content: files/en-us/web/css/css_basic_user_interface/index.md

  - Guide
  - Overview
  - Reference
spec-urls: 
  - https://drafts.csswg.org/css-ui/

>   - https://www.w3.org/TR/css-ui-3/

Reviewer: We never add old versions of the specs (unless the new is incremental, which is not the case here).

cc/ @Elchi3 or @wbamberg for confirmation
```suggestion
```

Author: one is level 3, the other is level 4

Author: I was just copying existing landing pages, like https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Text.

---

{{CSSRef}}

The **CSS basic user interface** module lets you define the rendering and functionality of features related to the user interface including outline properties, visual feedback to pointing device and keyboard, and altering the default appearance of UI widgets.

> Basic user interface properties can be used to improve user experience and accessibility by providing visual cues to elements that are being intereacted with, including styling mouse cursors and keyboard directional focus navigation, and styling caret cursors when an editable element has focus. The module provides for providing outlines to focused (or not) elements without impacting an element's [box model](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#what_is_the_css_box_model) dimensions and styling. This UI module also enables the styling of user interface controls.

Reviewer 2: ```suggestion
Basic user interface properties can be used to improve user experience and accessibility by providing visual cues to elements that are being interacted with, including styling mouse cursors and keyboard directional focus navigation, and styling caret cursors when an editable element has focus. The module provides for providing outlines to focused (or not) elements without impacting an element's [box model](/en-US/docs/Learn/CSS/Building_blocks/The_box_model#what_is_the_css_box_model) dimensions and styling. This UI module also enables the styling of user interface controls.
```

  caret-color: magenta;
  border: 1px solid #ccc;
}
:focus {
  outline: dashed magenta 3px;
  outline-offset: 10px;
}
* {
  accent-color: magenta;
}
div, fieldset {
  margin: 20px;
}

textarea:nth-of-type(1) {
  cursor: wait;
}
textarea:nth-of-type(2) {
  resize: none;
}
textarea:nth-of-type(3) {
  pointer-events: none;
}
```

{{ EmbedLiveSample('Basic_user_interface_in_action', "630", "300") }}

The CSS {{CSSxRef("outline")}} and {{CSSxRef("outline-offset")}} properties were used to provide feedback to users which element currently has focus. An {{CSSxRef("accent-color")}} provides a theme color to all the form controls. The caret that appears when the text is edit has the same color thanks to the {{CSSxRef("caret-color")}} property. These can all be considered UI improvements.

Some features harm usability. The {{CSSxRef("cursor")}} property was used to change cursors from the browser default which is confusing. The {{CSSxRef("resize")}} property prevents the second {{HTMLElement("textarea")}} from being resizable while the {{CSSxRef("pointer-events")}} property prevents the third `<textarea>` from receiving click events. It is still focusable using the keyboard.

> To see the code for this basic user interface sample, [view the source on Github](https://github.com/mdn/content/blob/main/files/en-us/web/css/CSS_Basic_User_Interface/index.md).

Reviewer: We shouldn't say to "view the source" on GitHub and link to a **Markdown** file with much more than the source. Readers will likely be lost, especially as beginners will be the most likely to check these features.

@wbamberg Could we use `{{EmbedGHLiveSample}}` so that:
- the code is only in one place
- the code is still displayed on this page

Author: We do not want to show the code on the page.

Reviewer 3: I agree `{{EmbedGHLiveSample}}` is a good fit here.

> We do not want to show the code on the page.

If I understand @teoli2003 correctly, by "this page" he means the page(s) linked to from "view the source", not the MDN page that the reader is seeing.

Author: Working on getting those pages up in css-examples, then i can update here.

To see the code for this basic user interface sample, [view the source on Github](https://github.com/mdn/content/blob/main/files/en-us/web/css/CSS_Basic_User_Interface/index.md).

## Reference

### Properties

- {{CSSxRef("appearance")}} {{Experimental_Inline}}

> - {{CSSxRef("box-sizing")}}

Reviewer: Box-sizing has been moved to its separate module between Levels 3 and 4.

How do we handle such cases? Should we add it (or the CSS module about it) to the _See also_ section. It seems this implies some relation between `box-sizing` and this module.

Author: it will be in the correct spec module landing page

> - {{CSSxRef("caret-color")}}
> - {{CSSxRef("caret-shape")}}
> - {{CSSxRef("caret")}} shorthand

Author: ```suggestion

- {{CSSxRef("caret")}} shorthand
  - {{CSSxRef("caret-color")}}
  - {{CSSxRef("caret-shape")}}
```

- {{CSSxRef("cursor")}}
- {{CSSxRef("nav-down")}} {{Experimental_Inline}}
- {{CSSxRef("nav-left")}} {{Experimental_Inline}}
- {{CSSxRef("nav-right")}} {{Experimental_Inline}}
- {{CSSxRef("nav-up")}} {{Experimental_Inline}}
- {{CSSxRef("outline-color")}}
- {{CSSxRef("outline-style")}}
- {{CSSxRef("outline-width")}}

> - {{CSSxRef("outline")}} shorthand

Reviewer: I don't understand the positioning of the shorthand.

Author: the value on line 107 is not part of the shorthand.

> - {{CSSxRef("outline-color")}}
> - {{CSSxRef("outline-style")}}
> - {{CSSxRef("outline-width")}}
> - {{CSSxRef("outline")}} shorthand

Reviewer: What about:
```suggestion
- {{CSSxRef("outline")}}, shorthand for:
  - {{CSSxRef("outline-color")}}
  - {{CSSxRef("outline-style")}}
  - {{CSSxRef("outline-width")}}
```

Author: I like it!

- {{CSSxRef("outline-offset")}}
- {{CSSxRef("pointer-events")}}
- {{CSSxRef("resize")}}
- {{CSSxRef("user-select")}}

> Previously, the {{CSSxRef("box-sizing")}} and {{CSSxRef("text-overflow")}} CSS properties were part of the basic user interface module, but were moved to the [CSS box sizing](/en-US/docs/Web/CSS/CSS_box_sizing) and [CSS overflow](/en-US/docs/Web/CSS/CSS_overflow) modules, respectively. The {{CSSxRef("ime-mode")}} property has been deprecated and should not be used. An `input-security` property is under consideration, but is the provenance of HTML.

Author: It is now neither here nor in the see also. But if it should be on the page, it should be here.

### Basic user interface in action

To view how basic user interface properties can alter the appearance of UI features, interact with the elements in this sample. Note that some features in this sample improve usability while others harm user experience.

{{EmbedGHLiveSample("css-examples/modules/basicUI.html", '100%', 320)}}

The CSS {{CSSxRef("outline")}} and {{CSSxRef("outline-offset")}} properties were used to provide feedback to users which element currently has focus. An {{CSSxRef("accent-color")}} provides a theme color to all the form controls. The caret that appears when the text is edit has the same color thanks to the {{CSSxRef("caret-color")}} property. These can all be considered UI improvements.

Some features harm usability. The {{CSSxRef("cursor")}} property was used to change cursors from the browser default which is confusing. The {{CSSxRef("resize")}} property prevents the second {{HTMLElement("textarea")}} from being resizable while the {{CSSxRef("pointer-events")}} property prevents the third `<textarea>` from receiving click events. It is still focusable using the keyboard.

> To see the code for this basic user interface sample, [view the source on Github](https://github.com/mdn/content/blob/main/files/en-us/web/css/CSS_Basic_User_Interface/index.md).

Author: ```suggestion
To see the code for this basic user interface sample, [view the source on Github](https://github.com/mdn/css-examples/blob/main/modules/basicUI.html).
```
